### PR TITLE
Add pattern to forward warpSingleLaneOp operands

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Add builder to create warpOp with operands and add a pattern to
propagate the operands when they are directly returned out of the region

### CODE REVIEW

File: include/Dialects/VectorExt/VectorExtOps.td [hunk 1]
```diff
@@ -148,8 +148,9 @@ def VectorExt_WarpSingleLaneOp : VectorExt_Op<"warp_execute_on_lane_0",
   let skipDefaultBuilders = 1;
   let builders = [
     OpBuilder<(ins "Value":$laneid)>,
-    // TODO: Add arguments support in the builder.
-    OpBuilder<(ins "TypeRange":$resultTypes, "Value":$laneid)>
+    OpBuilder<(ins "TypeRange":$resultTypes, "Value":$laneid)>,
+    OpBuilder<(ins "TypeRange":$resultTypes, "Value":$laneid,
+                  "ValueRange":$operands, "TypeRange":$argTypes)>
```
Comment (matthias-springer) [inline include/Dialects/VectorExt/VectorExtOps.td:153]: nit: indent

Comment (matthias-springer) [reply include/Dialects/VectorExt/VectorExtOps.td:153]: nit: I would call them `$args` instead of `$operands`

Comment (matthias-springer) [reply include/Dialects/VectorExt/VectorExtOps.td:153]: I would rename `argTypes` to `blockArgTypes` and add a comment that this cannot be inferred from `args` and is needed to distinguish between broadcasting/gathering from lanes (are these the right terms?) semantics.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp [hunk 2]
```diff
@@ -354,6 +355,37 @@ struct WarpOpDeadResult : public OpRewritePattern<WarpSingleLaneOp> {
...
+      Value warpOperand = warpOp.args()[arg.getArgNumber()];
+      if (result.use_empty() || result.getType() != warpOperand.getType())
+        continue;
+      valForwarded = warpOperand;
+      resultIndex = operand.getOperandNumber();
+      break;
+    }
+    if (!valForwarded)
+      return failure();
+    warpOp.getResult(resultIndex).replaceAllUsesWith(valForwarded);
```
Comment (matthias-springer) [inline lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:384]: Is this safe? I always thought we would have to go through the rewriter API or do something like `updateRootInplace` (which does not work with `replaceAllUsesWith`).

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp [hunk 3]
```diff
@@ -354,6 +355,37 @@ struct WarpOpDeadResult : public OpRewritePattern<WarpSingleLaneOp> {
...
+        warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+    Value valForwarded;
+    unsigned resultIndex;
+    for (OpOperand &operand : yield->getOpOperands()) {
+      auto arg = operand.get().dyn_cast<BlockArgument>();
+      if (!arg || arg.getOwner()->getParentOp() != warpOp.getOperation())
+        continue;
+      Value result = warpOp.getResult(operand.getOperandNumber());
+      Value warpOperand = warpOp.args()[arg.getArgNumber()];
+      if (result.use_empty() || result.getType() != warpOperand.getType())
```
Comment (matthias-springer) [inline lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:376]: Is this something that we support?

```mlir
%0 = vector_ext.warp_execute_on_lane_0(%laneid) args(%v0 : vector<1xf32>) -> (vector<1xf32>) {
 ^bb0(%arg0 : vector<1xf32>) :
  vector_ext.yield %arg0 : vector<1xf32>
}
```

(With the assumption that every lane has the same value.)

Or is this forbidden and we would just capture the `%v0` without going through `args`?

**Review by matthias-springer — APPROVED** [review]
Nice, I needed this for one of my changes.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp (hunk 3 above)
Comment (ThomasRaoux) [reply lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:376]: That's a good point. I think having warp_execute_on_lane_0 argument and the block argument have the same type should only be legal if the value is uniform. (or maybe we should disallow this completely). I would go with the assumption that this is correct to convert this example to `%v0` but we should think about it more and make sure the semantic is coherent with the rest.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:384]: This doesn't modify the root so I though we didn't need `updateRootInplace` here. I see some other examples upstream which seem to do something similar so I assumed it was correct:
https://github.com/llvm/llvm-project/blob/5d47332783d0421ba0da2a723354fb523eb85d0b/mlir/lib/Dialect/SCF/SCF.cpp#L1449

Although I'm not completely clear if this case is allowed

File: include/Dialects/VectorExt/VectorExtOps.td (hunk 1 above)
Comment (ThomasRaoux) [reply include/Dialects/VectorExt/VectorExtOps.td]: done.

**Pull request merged**